resources/src/main/java/org/robolectric/res/android/ResourceString.java

```diff
@@ -76,6 +78,22 @@ public static String decodeString(ByteBuffer buffer, int offset, Type type) {
     } else {
       length = characterCount * 2;
     }
+    if (type == Type.UTF8) {
+      // We should use DataInputStream to read UTF-8 and modified UTF-8 to avoid decoding
+      // problem. For example, the emoji/unicode is compressed with modified UTF-8, and
+      // normal String decoding will fail.
+      ByteBuffer bb = ByteBuffer.allocate(length + 2);
+      bb.putShort((short) length).put(buffer.array(), offset, length);
+      ByteArrayInputStream bis = new ByteArrayInputStream(bb.array());
+      DataInputStream dis = new DataInputStream(bis);
+      try {
+        return dis.readUTF();
+      } catch (IOException e) {
+        // It will fallback to normal String decoding if failed.
+        e.printStackTrace();
```

Comment:
Have you encountered a situation where DataInputStream fails to read the UTF-8 input but `new String` works?

Reply:
Sorry, I didn't encounter those problems. And I switched to use the `CESU8` decoder to re-implement it based on your provided `Apktool` solution. The new solution will use default decoder to decode, and use `CESU8` as fallback decoder for modified `UTF-8`.